File: 2024/Python/Day_3.py
import re


def part_1():
    total_result = 0
    # Parse data from txt file
    with open("Inputs/day_3_input.txt") as input_file:
        input_text = input_file.read()

    breaker = True

    while breaker:
        mul_index = input_text.find("mul(")
        if mul_index < 0:
            breaker = False
            break

        closing_paren_index = input_text[mul_index:].find(")") + mul_index + 1
        s = input_text[mul_index:closing_paren_index]

        if re.match(r"mul\(([0-9]|[1-9][0-9]|[1-9][0-9][0-9]),([0-9]|[1-9][0-9]|[1-9][0-9][0-9])\)", s):
            x = input_text[mul_index+4:closing_paren_index-1].split(',')
            total_result = total_result + (int(x[0]) * int(x[1]))
        input_text = input_text[mul_index+1:]
    return total_result

def part_2():
    total_result = 0
    # Parse data from txt file
    with open("inputs\\day_3_input.txt") as input_file:
        input_text = input_file.read()

    breaker = True
    do_dont_switch = True

    while breaker:
        mul_index = input_text.find("mul(")
        do_index = input_text.find("do")
        dont_index = input_text.find("don't")

        if mul_index < 0:
            breaker = False
            break

        if do_index == -1:
            do_index = 99999
        if dont_index == -1:
            dont_index = 99999

        if mul_index < do_index and mul_index < dont_index:
            if do_dont_switch:
                closing_paren_index = input_text[mul_index:].find(")") + mul_index + 1
                s = input_text[mul_index:closing_paren_index]

                if re.match(r"mul\(([0-9]|[1-9][0-9]|[1-9][0-9][0-9]),([0-9]|[1-9][0-9]|[1-9][0-9][0-9])\)", s):
                    x = input_text[mul_index + 4:closing_paren_index - 1].split(',')
                    total_result = total_result + (int(x[0]) * int(x[1]))
                input_text = input_text[mul_index + 1:]
            else:
                input_text = input_text[mul_index + 1:]
        elif do_index < mul_index and do_index < dont_index:
            do_dont_switch = True
            input_text = input_text[do_index + 1:]
        elif dont_index < mul_index and dont_index < do_index or (dont_index == do_index):
            do_dont_switch = False
            input_text = input_text[dont_index + 1:]
    return total_result


if __name__ == '__main__':
    print("Total Result: " + str(part_1()))
    print("Total Result: " + str(part_2()))

File: 2024/Rust/aoc_2024/src/day_3.rs
use std::fs;
use regex::Regex;

pub fn day_3_main(){
    part_1();
    part_2();
}

fn part_1(){
    //Ingest data
    let file_name = "inputs/day_3_input.txt";
    let contents:String = fs::read_to_string(file_name)
        .expect("Failed to read File");

    let mut breaker:bool = true;
    let mut sumr:i32 = 0;

    while breaker{
        let re = Regex::new(r"mul\(([0-9]|[1-9][0-9]|[1-9][0-9][0-9]),([0-9]|[1-9][0-9]|[1-9][0-9][0-9])\)").unwrap();
        let mut_index:usize = contents.find("mut");

        let n_string:String = contents.chars().skip(mut_index).collect();




        /*
        // TODO: COnvert this down
        mul_index = input_text.find("mul(")
        if mul_index < 0:
            breaker = False
            break

        closing_paren_index = input_text[mul_index:].find(")") + mul_index + 1
        s = input_text[mul_index:closing_paren_index]

        if re.match(r"mul\(([0-9]|[1-9][0-9]|[1-9][0-9][0-9]),([0-9]|[1-9][0-9]|[1-9][0-9][0-9])\)", s):
            x = input_text[mul_index+4:closing_paren_index-1].split(',')
            total_result = total_result + (int(x[0]) * int(x[1]))
        input_text = input_text[mul_index+1:]
        */

        breaker = false;
    }

    println!("Total multiple from Part 1: {}", sumr);
}

fn part_2(){
    //Ingest data
    let file_name = "inputs/day_2_input.txt";
    let contents = fs::read_to_string(file_name)
        .expect("Failed to read File");

    let mut breaker:bool = true;
    let mut sumr:i32 = 0;

    println!("Total multiple from Part 2: {}", sumr);
}
